control-plane/api-gateway/controllers/gatewayclass_controller_test.go

```diff
@@ -0,0 +1,176 @@
+		expectedResult     ctrl.Result
+		expectedError      error
+		expectedFinalizers []string
+	}{
+		"successful reconcile with no change": {
+			gatewayClass: &gwv1beta1.GatewayClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      name,
+					Finalizers: []string{
+						GatewayClassFinalizer,
+					},
+				},
+				Spec: gwv1beta1.GatewayClassSpec{
+					ControllerName: GatewayClassControllerName,
+				},
+			},
+
+			expectedResult:     ctrl.Result{},
+			expectedError:      nil,
+			expectedFinalizers: []string{GatewayClassFinalizer},
+		},
+		"successful reconcile that adds finalizer": {
+			gatewayClass: &gwv1beta1.GatewayClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace:  namespace,
+					Name:       name,
+					Finalizers: []string{},
+				},
+				Spec: gwv1beta1.GatewayClassSpec{
+					ControllerName: GatewayClassControllerName,
+				},
+			},
+			expectedResult:     ctrl.Result{Requeue: true},
+			expectedError:      nil,
+			expectedFinalizers: []string{GatewayClassFinalizer},
+		},
+		"attempt to reconcile a GatewayClass with a different controller name": {
+			gatewayClass: &gwv1beta1.GatewayClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace:  namespace,
+					Name:       name,
+					Finalizers: []string{},
+				},
+				Spec: gwv1beta1.GatewayClassSpec{
+					ControllerName: "foo",
+				},
+			},
+			expectedResult: ctrl.Result{},
+			expectedError:  nil,
+		},
+		"attempt to reconcile a non-existent object": {
+			k8sObjects:     []runtime.Object{},
+			expectedResult: ctrl.Result{},
+			expectedError:  nil,
+		},
+		"attempt to remove a GatewayClass that is not in use": {
+			gatewayClass: &gwv1beta1.GatewayClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      name,
+					Finalizers: []string{
+						GatewayClassFinalizer,
+					},
+					DeletionTimestamp: &deletionTimestamp,
+				},
+				Spec: gwv1beta1.GatewayClassSpec{
+					ControllerName: GatewayClassControllerName,
+				},
+			},
+			expectedResult:     ctrl.Result{},
+			expectedError:      nil,
+			expectedFinalizers: []string{},
+		},
+		"attempt to remove a GatewayClass that is in use": {
+			gatewayClass: &gwv1beta1.GatewayClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      name,
+					Finalizers: []string{
+						GatewayClassFinalizer,
+					},
+					DeletionTimestamp: &deletionTimestamp,
+				},
+				Spec: gwv1beta1.GatewayClassSpec{
+					ControllerName: GatewayClassControllerName,
+				},
+			},
+			k8sObjects: []runtime.Object{
+				&gwv1beta1.Gateway{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "test-gateway",
+					},
+					Spec: gwv1beta1.GatewaySpec{
+						GatewayClassName: v1beta1.ObjectName(name),
+					},
+				},
+			},
+			expectedResult:     ctrl.Result{RequeueAfter: 10 * time.Second},
+			expectedError:      nil,
+			expectedFinalizers: []string{GatewayClassFinalizer},
+		},
```

Comment:
Can we add a couple tests for the `GatewayClassConfig` validation logic? Also for checking that the status is set appropriately?